src/CarbonAware.Aggregators/test/ForecastAggregatorTests.cs

```diff
@@ -0,0 +1,277 @@
+            .ReturnsAsync(forecast);
+
+        var parameters = new CarbonAwareParametersBaseDTO()
+        {
+            MultipleLocations = new string[] { "westus" },
+            Start = start,
+            End = end
+        };
+
+        // Act
+        var results = await this.Aggregator.GetCurrentForecastDataAsync(parameters);
+
+        // Assert
+        var forecastResult = results.First();
+        Assert.AreEqual(expectedStart, forecastResult.DataStartAt);
+        Assert.AreEqual(expectedEnd, forecastResult.DataEndAt);
+    }
+
+    [Test]
+    public void TestGetCurrentForecastDataAsync_NoLocation()
+    {
+        Assert.ThrowsAsync<ArgumentException>(
+            async () => await this.Aggregator.GetCurrentForecastDataAsync(new CarbonAwareParametersBaseDTO()));
+    }
+
+    [TestCase("2022-01-01T00:00:00Z", "2022-01-01T00:20:00Z", ExpectedResult = 4, TestName = "Start and end time match")]
+    [TestCase("2022-01-01T00:02:00Z", "2022-01-01T00:12:00Z", ExpectedResult = 3, TestName = "Start and end match midway through a datapoint")]
+    public async Task<int> TestGetCurrentForecastDataAsync_FiltersDate(DateTimeOffset? start, DateTimeOffset? end)
+    {
+        this.ForecastDataSource.Setup(x => x.GetCurrentCarbonIntensityForecastAsync(It.IsAny<Location>()))
+            .ReturnsAsync(TestData.GetForecast("2022-01-01T00:00:00Z"));
+        var parameters = new CarbonAwareParametersBaseDTO()
+        {
+            MultipleLocations = new string[] { "westus" },
+            Start = start,
+            End = end
+        };
+
+        var results = await this.Aggregator.GetCurrentForecastDataAsync(parameters);
+        var forecastData = results.First().ForecastData;
+
+        return forecastData.Count();
+    }
+
+    [TestCase("2022-01-01T00:00:00Z", "2022-01-01T00:20:00Z", TestName = "Full data set")]
+    [TestCase("2022-01-01T00:05:00Z", "2022-01-01T00:20:00Z", TestName = "Data set minus first lowest datapoint")]
+    public async Task TestGetCurrentForecastDataAsync_OptimalDataPoint(DateTimeOffset? start, DateTimeOffset? end)
+    {
+        this.ForecastDataSource.Setup(x => x.GetCurrentCarbonIntensityForecastAsync(It.IsAny<Location>()))
+            .ReturnsAsync(TestData.GetForecast("2022-01-01T00:00:00Z"));
+        var parameters = new CarbonAwareParametersBaseDTO()
+        {
+            MultipleLocations = new string[] { "westus" },
+            Start = start,
+            End = end
+        };
+
+        var results = await this.Aggregator.GetCurrentForecastDataAsync(parameters);
+        var forecast = results.First();
+        var firstDataPoint = forecast.ForecastData.First();
+        var optimalDataPoint = forecast.OptimalDataPoints.First();
+
+        Assert.AreEqual(firstDataPoint, optimalDataPoint);
+    }
+
+    [Test]
+    public async Task TestGetCurrentForecastDataAsync_RollingAverageWithSpecifiedWindow()
```

Comment:
Given we've seen some instances where either too few data points or "zero'ed" out data points cause us issues, should we add some test cases in here to account for those scenarios?

Reply:
The data source layer has the logic to error out in case of too few data points. We have added tests there to cover such scenarios